File: src/com/colacelli/ircbot/plugins/apertiumtranslate/ApertiumTranslatePlugin.java
package com.colacelli.ircbot.plugins.apertiumtranslate;

import com.colacelli.ircbot.IRCBot;
import com.colacelli.ircbot.Plugin;
import com.colacelli.ircbot.listeners.OnChannelCommandListener;
import com.colacelli.ircbot.plugins.access.IRCBotAccess;
import com.colacelli.ircbot.plugins.help.PluginHelp;
import com.colacelli.ircbot.plugins.help.PluginHelper;
import com.colacelli.irclib.connection.Connection;
import com.colacelli.irclib.messages.ChannelMessage;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.Scanner;

public class ApertiumTranslatePlugin implements Plugin {
    private static final String APERTIUM_URL = "https://www.apertium.org/apy/translate?q=TEXT&langpair=LOCALES";
    private static final String APERTIUM_JSON_DATA = "responseData";
    private static final String APERTIUM_JSON_TRANSLATION = "translatedText";

    private OnChannelCommandListener listener;

    public ApertiumTranslatePlugin() {
        listener = new OnChannelCommandListener() {
            @Override
            public String channelCommand() {
                return ".translate";
            }

            @Override
            public void onChannelCommand(Connection connection, ChannelMessage message, String command, String... args) {
                if (args != null && args.length > 2) {
                    String localeA = args[0];
                    String localeB = args[1];
                    StringBuilder text = new StringBuilder();

                    text.append(args[2]);

                    for (int i = 3; i < args.length; i++) {
                        text.append(" ");
                        text.append(args[i]);
                    }

                    // FIXME: Move it into a thread
                    String translation = getTranslation(localeA, localeB, text.toString());

                    if (translation == null || translation.isEmpty()) {
                        translation = "Translation not found!";
                    }

                    ChannelMessage.Builder builder = new ChannelMessage.Builder();
                    builder
                            .setChannel(message.getChannel())
                            .setSender(connection.getUser())
                            .setText("[" + localeA + "] " + text + " ~ " + "[" + localeB + "] " + translation);

                    connection.send(builder.build());
                }
            }
        };
    }

    @Override
    public String getName() {
        return "APERTIUM_TRANSLATE";
    }

    @Override
    public void onLoad(IRCBot bot) {
        bot.addListener(listener);

        PluginHelper.getInstance().addHelp(new PluginHelp(
                ".translate",
                IRCBotAccess.USER_LEVEL,
                "Translate text from locale1 to locale2 using Apertium (https://apertium.org)",
                "locale1",
                "locale2"));
    }

    @Override
    public void onUnload(IRCBot bot) {
        bot.removeListener(".translate");
        PluginHelper.getInstance().removeHelp(".translate");
    }

    private String getTranslation(String localeA, String localeB, String text) {
        String url = APERTIUM_URL
                .replace("LOCALES", localeA + "|" + localeB)
                .replace("TEXT", text.replaceAll(" ", "%20"));

        String translation = null;

        try {
            StringBuilder jsonText = new StringBuilder();

            InputStream response = new URL(url).openStream();
            Scanner scanner = new Scanner(response).useDelimiter("\\A");
            while (scanner.hasNext()) {
                jsonText.append(scanner.next());
            }

            JSONObject json = (JSONObject) new JSONParser().parse(jsonText.toString());
            json = (JSONObject) json.get(APERTIUM_JSON_DATA);

            translation = (String) json.get(APERTIUM_JSON_TRANSLATION);
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }

        return translation;
    }
}

File: src/com/colacelli/ircbot/plugins/apertiumtranslate/ApertiumTranslatePlugin.kt
package com.colacelli.ircbot.plugins.apertiumtranslate

import com.colacelli.ircbot.IRCBot
import com.colacelli.ircbot.Plugin
import com.colacelli.ircbot.listeners.OnChannelCommandListener
import com.colacelli.ircbot.plugins.access.IRCBotAccess
import com.colacelli.ircbot.plugins.help.PluginHelp
import com.colacelli.ircbot.plugins.help.PluginHelper
import com.colacelli.irclib.connection.Connection
import com.colacelli.irclib.messages.ChannelMessage
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import java.net.URL
import java.util.*

class ApertiumTranslatePlugin : Plugin {
    private val listener = object : OnChannelCommandListener {
        override val commands: Array<String>
            get() = arrayOf(".translate", ".trans")

            override fun onChannelCommand(connection: Connection, message: ChannelMessage, command: String, args: Array<String>) {
                if (args.size > 2) {
                    val localeA = args[0]
                    val localeB = args[1]
                    var text = args[2]

                    for (i in 3 until args.size) {
                        text += " ${args[i]}"
                    }

                    val apertiumTranslate = ApertiumTranslate(localeA, localeB, text)

                    apertiumTranslate.addListener(object : OnApertiumTranslateResultListener {
                        override fun onSuccess(translation: ApertiumTranslation) {
                            connection.send(ChannelMessage(
                                    message.channel,
                                    "[${translation.localeA}] ${translation.text} ~ [${translation.localeB}] ${translation.translation}",
                                    connection.user
                            ))
                        }

                        override fun onError() {
                            connection.send(ChannelMessage(message.channel, "Translation not found!", connection.user))
                        }
                    })

                    val worker = Thread(apertiumTranslate)
                    worker.name = "apertium_translate"
                    worker.start()
                }
            }

        }

    override fun getName(): String {
        return "apertium_translate"
    }

    override fun onLoad(bot: IRCBot) {
        bot.addListener(listener)

        PluginHelper.instance.addHelp(PluginHelp(
                ".translate",
                IRCBotAccess.Level.USER,
                "Translate text from locale1 to locale2 using Apertium (https://apertium.org)",
                "locale1",
                "locale2"))
    }

    override fun onUnload(bot: IRCBot) {
        bot.removeListener(".translate")
        PluginHelper.instance.removeHelp(".translate")
    }

    private class ApertiumTranslate(private val localeA: String, private val localeB: String, private val text: String) : Runnable {
        var listeners = ArrayList<OnApertiumTranslateResultListener>()

        companion object {
            const val APERTIUM_URL = "https://www.apertium.org/apy/translate?q=TEXT&langpair=LOCALES"
        }

        override fun run() {
            val url = APERTIUM_URL
                .replace("LOCALES", "$localeA|$localeB")
                .replace("TEXT", text)

            val stream = URL(url).openStream()
            val scanner = Scanner(stream).useDelimiter("\\A")

            var json = ""
            while (scanner.hasNext()) {
                json += scanner.next()
            }

            val gson = Gson()
            val response = gson.fromJson(json, ApertiumTranslationResponse::class.java)

            listeners.forEach {
                if (response.status == 200) {
                    // FIXME: Remove it from the forEach, and try to be assigned by Gson
                    val translation = response.data
                    translation.localeA = localeA
                    translation.localeB = localeB
                    translation.text = text

                    it.onSuccess(translation)
                } else {
                    it.onError()
                }
            }
        }

        fun addListener(listener : OnApertiumTranslateResultListener) {
            listeners.add(listener)
        }
    }

    private class ApertiumTranslationResponse(@SerializedName("responseStatus") val status: Int, @SerializedName("responseData") val data: ApertiumTranslation)
    class ApertiumTranslation(var localeA: String, var localeB: String, var text: String, @SerializedName("translatedText") val translation: String)
}